{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7cdeaf74",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from src.files.blast import BLASTJsonFile\n",
    "import re\n",
    "import os \n",
    "import pandas as pd \n",
    "from tqdm import tqdm\n",
    "\n",
    "Entrez.email = 'prichter@berkeley.edu'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257828e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ids(result):\n",
    "    ids = [match.group(1) for match in re.finditer(r'<Id>(\\d+)</Id>', result)]\n",
    "    return ids\n",
    "\n",
    "def get_coordinates(result): # Takes a protein fetch result as input.\n",
    "    pattern = r'<GBQualifier_name>coded_by</GBQualifier_name>\\s+<GBQualifier_value>(.+)</GBQualifier_value>'\n",
    "    match = re.search(pattern, result, flags=re.MULTILINE)\n",
    "    return match.group(1) if (match is not None) else 'none'\n",
    "\n",
    "def get_taxonomy(result): # Takes a nuccore fetch result as input.\n",
    "    pattern = r'<GBSeq_taxonomy>(.+)</GBSeq_taxonomy>'\n",
    "    match = re.search(pattern, result, flags=re.DOTALL)\n",
    "    return match.group(1) if (match is not None) else 'none'\n",
    "\n",
    "def get_organism(result): # Takes a nuccore fetch result as input.\n",
    "    pattern = r'<GBSeq_organism>(.+)</GBSeq_organism>'\n",
    "    match = re.search(pattern, result, flags=re.DOTALL)\n",
    "    return match.group(1) if (match is not None) else 'none'\n",
    "\n",
    "# def get_nuccore_accession(result):\n",
    "#     pattern = r'<GBSeq_accession-version>(.+)</GBSeq_accession-version>'\n",
    "#     match = re.search(pattern, result, flags=re.DOTALL)\n",
    "#     return match.group(1) if (match is not None) else 'none'\n",
    "\n",
    "def get_nuccore_accession(result): # Takes a protein fetch result as input. \n",
    "    pattern = r'<GBSeq_source-db>accession (.+)</GBSeq_source-db'\n",
    "    match = re.search(pattern, result, flags=re.DOTALL)\n",
    "    return match.group(1) if (match is not None) else 'none'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1ba942",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_fasta(nuccore_id:str, fn_dir=None):\n",
    "    fn_path = os.path.join(fn_dir, f'{nuccore_id}.fn')\n",
    "    if not os.path.exists(fn_path):\n",
    "        result = Entrez.efetch(db='nuccore', id=nuccore_id, rettype='fasta').read()\n",
    "        with open(fn_path, 'w') as f:\n",
    "            f.write(result)\n",
    "    return fn_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "19bc0392",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. BLAST hits: 549\n"
     ]
    }
   ],
   "source": [
    "blast_df = BLASTJsonFile('../data/arf1-2.json').to_df()\n",
    "blast_df = blast_df.drop_duplicates('subject_id')\n",
    "print('Num. BLAST hits:', len(blast_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f06aeaf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_nr_protein_info(id_, fn_dir='../data/ncbi/nucleotides'):\n",
    "    assert id_.startswith('WP'), f'download_nr_protein_info: Expected a protein accession beginning with WP, but got {id_}'\n",
    "    result = Entrez.efetch(db='ipg', id=id_, rettype='html').read().decode('utf-8')\n",
    "    \n",
    "    info = list()\n",
    "    pattern = r'<CDS  accver=\"(.+)\" start=\"(\\d+)\" stop=\"(\\d+)\" strand=\"(.+)\" taxid=\"(.+)\" org=\"([^\"]+)\".+</CDSList></Protein>'\n",
    "    for match in re.finditer(pattern, result):\n",
    "        info_ = dict()\n",
    "        info_['nuccore_id'] = match.group(1)\n",
    "        info_['start'] = match.group(2)\n",
    "        info_['stop'] = match.group(3)\n",
    "        info_['strand'] = match.group(4)\n",
    "        info_['organism'] = match.group(6)\n",
    "        info_['html'] = match.group(0)\n",
    "        info.append(info_)\n",
    "\n",
    "        download_fasta\n",
    "\n",
    "    return info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eadd23db",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_protein_info(id_, fn_dir:str='../data/ncbi/nucleotides'):\n",
    "    # TODO: The non-redundant protein sequences need to be handled differently.  \n",
    "    info = {'id':id_}\n",
    "    \n",
    "    result = Entrez.efetch(db='protein', id=id_, rettype='html').read().decode('utf-8')\n",
    "    nuccore_id = get_nuccore_accession(result)\n",
    "    if nuccore_id == 'none':\n",
    "        return info\n",
    "\n",
    "    info['coordinates'] = get_coordinates(result)\n",
    "    info['nuccore_id'] = nuccore_id\n",
    "\n",
    "    result = Entrez.efetch(db='nuccore', id=nuccore_id, rettype='html').read().decode('utf-8')\n",
    "\n",
    "    info['organism'] = get_organism(result)\n",
    "    info['taxonomy'] = get_taxonomy(result)\n",
    "    info['fn_path'] = download_fasta(nuccore_id, fn_dir=fn_dir)\n",
    "    \n",
    "    return info\n",
    "\n",
    "# arf12_df = list()\n",
    "# for id_ in tqdm(blast_df.subject_id, desc='Downloading aRF1-2 sequences...'):\n",
    "#     arf12_df.append(download_protein_info(id_))\n",
    "# arf12_df = pd.DataFrame(arf12_df).set_index('id')\n",
    "# arf12_df.to_csv('../data/arf1-2_blast_hits.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4dd64a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"  ?>\n",
      "<IPGReportSet xmlns:xsi=\"https://www.w3.org/2001/XMLSchema-instance\" xsi:noNamespaceSchemaLocation=\"https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd\">\n",
      "\n",
      "<IPGReport  ipg=\"501664124\" product_acc=\"WP_229389239.1\">\n",
      "<Product  accver=\"WP_229389239.1\" name=\"peptide chain release factor aRF-1\" taxid=\"2605639\" slen=\"415\" org=\"Methanosarcina sp. DH2\" kingdom_taxid=\"2157\" kingdom=\"Archaea\"/>\n",
      "<ProteinList>\n",
      "<Protein  accver=\"WP_229389239.1\" source=\"RefSeq\" name=\"peptide chain release factor aRF-1\" taxid=\"2605639\" org=\"Methanosarcina sp. DH2\" kingdom_taxid=\"2157\" kingdom=\"Archaea\" priority=\"0\">\n",
      "<CDSList>\n",
      "<CDS  accver=\"NZ_WJBI01000002.1\" start=\"378548\" stop=\"379795\" strand=\"-\" taxid=\"2605639\" org=\"Methanosarcina sp. DH2\" kingdom_taxid=\"2157\" kingdom=\"Archaea\" strain=\"DH2\" assembly=\"GCF_020804225.1\"/></CDSList></Protein>\n",
      "<Protein  accver=\"MCC4769673.1\" source=\"INSDC\" name=\"peptide chain release factor 1\" taxid=\"2605639\" org=\"Methanosarcina sp. DH2\" kingdom_taxid=\"2157\" kingdom=\"Archaea\" priority=\"60\">\n",
      "<CDSList>\n",
      "<CDS  accver=\"WJBI01000002.1\" start=\"378548\" stop=\"379795\" strand=\"-\" taxid=\"2605639\" org=\"Methanosarcina sp. DH2\" kingdom_taxid=\"2157\" kingdom=\"Archaea\" strain=\"DH2\" assembly=\"GCA_020804225.1\"/></CDSList></Protein></ProteinList>\n",
      "<Statistics  prot_count=\"2\" nuc_count=\"2\" assmb_count=\"2\"/></IPGReport></IPGReportSet>\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'nuccore_id': 'NZ_WJBI01000002.1',\n",
       "  'start': '378548',\n",
       "  'stop': 'NZ_WJBI01000002.1',\n",
       "  'strand': 'NZ_WJBI01000002.1',\n",
       "  'organism': 'Methanosarcina sp. DH2'},\n",
       " {'nuccore_id': 'WJBI01000002.1',\n",
       "  'start': '378548',\n",
       "  'stop': 'WJBI01000002.1',\n",
       "  'strand': 'WJBI01000002.1',\n",
       "  'organism': 'Methanosarcina sp. DH2'}]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print(Entrez.efetch(db='protein', id='2154289639', rettype='html').read().decode('utf-8'))\n",
    "# print(Entrez.esearch(db='protein', term='2154289639', rettype='html').read().decode('utf-8'))\n",
    "# print(Entrez.efetch(db='protein', term='2154289639', rettype='html').read().decode('utf-8'))\n",
    "# print(Entrez.elink(dbfrom='protein', db='protein', id='WP_229389239').read().decode('utf-8')) # 2154289639\n",
    "print(Entrez.efetch(db='ipg', id='WP_229389239', rettype='html').read().decode('utf-8')) # 2154289639\n",
    "# print(Entrez.efetch(db='nuccore', id='2154289639', rettype='html').read().decode('utf-8'))\n",
    "\n",
    "get_ipg_info()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c93c33a",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'arf12_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m wp_ids = [id_ \u001b[38;5;28;01mfor\u001b[39;00m id_ \u001b[38;5;129;01min\u001b[39;00m \u001b[43marf12_df\u001b[49m.index \u001b[38;5;28;01mif\u001b[39;00m id_.startswith(\u001b[33m'\u001b[39m\u001b[33mWP\u001b[39m\u001b[33m'\u001b[39m)]\n",
      "\u001b[31mNameError\u001b[39m: name 'arf12_df' is not defined"
     ]
    }
   ],
   "source": [
    "wp_ids = [id_ for id_ in arf12_df.index if id_.startswith('WP')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a00f318",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coordinates</th>\n",
       "      <th>nuccore_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>fn_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP_229389239</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_048180500</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_048140256</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_300104941</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_322225704</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_292518519</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_301664146</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_042705156</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_253488479</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_211530351</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             coordinates nuccore_id organism taxonomy fn_path\n",
       "id                                                           \n",
       "WP_229389239         NaN        NaN      NaN      NaN     NaN\n",
       "WP_048180500         NaN        NaN      NaN      NaN     NaN\n",
       "WP_048140256         NaN        NaN      NaN      NaN     NaN\n",
       "WP_300104941         NaN        NaN      NaN      NaN     NaN\n",
       "WP_322225704         NaN        NaN      NaN      NaN     NaN\n",
       "...                  ...        ...      ...      ...     ...\n",
       "WP_292518519         NaN        NaN      NaN      NaN     NaN\n",
       "WP_301664146         NaN        NaN      NaN      NaN     NaN\n",
       "WP_042705156         NaN        NaN      NaN      NaN     NaN\n",
       "WP_253488479         NaN        NaN      NaN      NaN     NaN\n",
       "WP_211530351         NaN        NaN      NaN      NaN     NaN\n",
       "\n",
       "[216 rows x 5 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arf12_df.fn_path.dropna()\n",
    "arf12_df[arf12_df.fn_path.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d942ae01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51970bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def download_protein_info(id_, fn_dir:str='../data/ncbi/nucleotides'):\n",
    "    \n",
    "#     info = list()\n",
    "    \n",
    "    # result = Entrez.esearch(db='protein', term=id_, rettype='html').read().decode('utf-8')\n",
    "    # protein_ids = get_ids(result)\n",
    "    # if len(protein_ids) == 0:\n",
    "    #     print(f'download_protein_info: No results for {id_}')\n",
    "    #     return info\n",
    "    \n",
    "#     for protein_id in protein_ids:\n",
    "#         info = {'id':id_, 'numerical_protein_id':protein_id}\n",
    "    \n",
    "#         result = Entrez.efetch(db='protein', id=protein_id, rettype='html').read().decode('utf-8')\n",
    "#         info['nuccore_id'], info['coordinates'] = get_coordinates(result)\n",
    "        \n",
    "#         result = Entrez.elink(dbfrom='protein', db='nuccore', id=protein_id, rettype='html').read().decode('utf-8')\n",
    "#         nuccore_id = get_ids(result)\n",
    "#         if len(nuccore_id) == 0:\n",
    "#             info.append(info)\n",
    "#             continue \n",
    "#         else:\n",
    "#             nuccore_id = nuccore_id[1] # First ID tag is just the ID itself. \n",
    "\n",
    "#         result = Entrez.efetch(db='nuccore', id=nuccore_id, rettype='html').read().decode('utf-8')\n",
    "\n",
    "#         info['organism'] = get_organism(result)\n",
    "#         info['taxonomy'] = get_taxonomy(result)\n",
    "#         # info['nuccore_id'] = get_nuccore_accession(result)\n",
    "#         info['numerical_nuccore_id'] = nuccore_id\n",
    "\n",
    "#         fn_path = os.path.join(fn_dir, f'{nuccore_id}.fn')\n",
    "#         if not os.path.exists(fn_path):\n",
    "#             result = Entrez.efetch(db='nuccore', id=nuccore_id, rettype='fasta').read()\n",
    "#             with open(fn_path, 'w') as f:\n",
    "#                 f.write(result)\n",
    "#             info['fn_path'] = fn_path\n",
    "#         info.append(info)\n",
    "    \n",
    "#     return info\n",
    "\n",
    "# arf12_df = list()\n",
    "# for id_ in tqdm(blast_df.subject_id, desc='Downloading aRF1-2 sequences...'):\n",
    "#     arf12_df += download_protein_info(id_)\n",
    "# arf12_df = pd.DataFrame(arf12_df).set_index('id')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyrrolysine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
